builtin/k8s/platform.go

```diff
@@ -476,132 +499,230 @@ func (p *Platform) resourceDeploymentCreate(
+				},
+			}
+		} else {
+			// If no probe path is defined, assume app will bind to default TCP port
+			// TODO: handle apps that aren't socket listeners
+			handler = corev1.Handler{
 				TCPSocket: &corev1.TCPSocketAction{
 					Port: intstr.FromInt(defaultPort),
 				},
-			},
+			}
+		}
+
+		container.LivenessProbe = &corev1.Probe{
+			Handler:             handler,
 			InitialDelaySeconds: initialDelaySeconds,
 			TimeoutSeconds:      timeoutSeconds,
 			FailureThreshold:    failureThreshold,
-		},
-		ReadinessProbe: &corev1.Probe{
-			Handler: corev1.Handler{
-				TCPSocket: &corev1.TCPSocketAction{
-					Port: intstr.FromInt(defaultPort),
-				},
-			},
+		}
+		container.ReadinessProbe = &corev1.Probe{
+			Handler:             handler,
 			InitialDelaySeconds: initialDelaySeconds,
 			TimeoutSeconds:      timeoutSeconds,
-		},
-		Env:       env,
-		Resources: resourceRequirements,
+		}
+	}
+
+	return &container, nil
+}
+
+// resourceDeploymentCreate creates the Kubernetes deployment.
+func (p *Platform) resourceDeploymentCreate(
+	ctx context.Context,
+	log hclog.Logger,
+	src *component.Source,
+	img *docker.Image,
+	deployConfig *component.DeploymentConfig,
+	ui terminal.UI,
+	result *Deployment,
+	state *Resource_Deployment,
+	csinfo *clientsetInfo,
+	sg terminal.StepGroup,
+) error {
+	// Prepare our namespace and override if set.
+	ns := csinfo.Namespace
+	if p.config.Namespace != "" {
+		ns = p.config.Namespace
+	}
+
+	step := sg.Add("")
+	defer func() { step.Abort() }()
+	step.Update("Kubernetes client connected to %s with namespace %s", csinfo.Config.Host, ns)
+	step.Done()
+
+	step = sg.Add("Preparing deployment...")
+
+	clientSet := csinfo.Clientset
+	deployClient := clientSet.AppsV1().Deployments(ns)
+
+	// Determine if we have a deployment that we manage already
+	create := false
+	deployment, err := deployClient.Get(ctx, result.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		deployment = result.newDeployment(result.Name)
+		create = true
+		err = nil
+	}
+	if err != nil {
+		return err
 	}
 
+	var overlayTarget *Container
 	if p.config.Pod != nil && p.config.Pod.Container != nil {
-		containerCfg := p.config.Pod.Container
-		if containerCfg.Command != nil {
-			container.Command = *containerCfg.Command
-		}
+		overlayTarget = p.config.Pod.Container
+	} else {
+		overlayTarget = &Container{}
+	}
 
-		if containerCfg.Args != nil {
-			container.Args = *containerCfg.Args
-		}
+	appContainerSpec, err := overlayTopLevelProperties(p.config, overlayTarget)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "Failed to parse container config: %s", err)
 	}
 
-	// Update the deployment with our spec
-	deployment.Spec.Template.Spec = corev1.PodSpec{
-		Containers: []corev1.Container{container},
+	// App container must have some kind of port
+	if len(appContainerSpec.Ports) == 0 {
+		ui.Output(fmt.Sprintf("No ports defined - defaulting to http on port %d", DefaultServicePort), terminal.WithWarningStyle())
```

Comment:
I think ui.Output should also let you format without `fmt.Sprintf` if I remember right

```suggestion
		ui.Output("No ports defined - defaulting to http on port %d", DefaultServicePort, terminal.WithWarningStyle())
```